### PR TITLE
Add support for 1.19 specific features.

### LINKED ISSUE #231: [1.19] Sculk spreading isn't logged
Spreading sculk with a catalyst currently isn't logged, making it impossible to rollback players that have spread sculk onto blocks.

image

### PR DESCRIPTION
**Pull request by Gungadean**
- Add config option and logging for sculk catalysts.
- Add support for new logging and restoring new Goat horn functionalities.
resolves #231 , #235

### CODE REVIEW

File: src/main/java/net/coreprotect/bukkit/Bukkit_v1_19.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.bukkit.entity.*;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+
+public class Bukkit_v1_19 extends Bukkit_v1_18 implements BukkitInterface {
+
+    public Bukkit_v1_19() {
+        BlockGroup.SCULK = new HashSet<>(Arrays.asList(Material.SCULK, Material.SCULK_VEIN));
```
Comment (Johnibur) [inline src/main/java/net/coreprotect/bukkit/Bukkit_v1_19.java:14]: You are missing a few materials here. Sculk growth has also a change to generate a sculk sensor and a sculk shrieker.

File: src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java [hunk 2]
```diff
@@ -27,20 +27,18 @@ protected void onBlockSpread(BlockSpreadEvent event) {
          * block-change: true
          *
          */
+        if (event.isCancelled()) {
+            return;
+        }
 
-        if (!event.isCancelled() && Config.getConfig(event.getBlock().getWorld()).VINE_GROWTH) {
```
Comment (Johnibur) [inline src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java:31]: Reviewing the logic seems necessary here, but it's a completely different topic from 1.19.

File: src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java [hunk 3]
```diff
@@ -74,6 +72,25 @@ else if (type.equals(Material.BAMBOO)) {
                 Queue.queueBlockPlaceDelayed("#bamboo", block.getLocation(), type, null, block.getState(), 0);
             }
         }
+
+        if(Config.getConfig(event.getBlock().getWorld()).SCULK_SPREAD && BlockGroup.SCULK.contains(type)) {
+            if(checkCacheData(block, type)) {
+                return;
+            }
+
+            queueBlockPlace("#sculkcatalyst", block.getState(), block.getType(), block.getState(), type, -1, 0, blockstate.getBlockData().getAsString());
```
Comment (Johnibur) [inline src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java:81]: I would simply call that `#sculk` instead.

File: src/main/java/net/coreprotect/paper/PaperAdapter.java [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package net.coreprotect.paper;
 
+import com.destroystokyo.paper.entity.villager.Reputation;
```
[the remark this answers is not in the document]
Comment (Gungadean) [reply src/main/java/net/coreprotect/paper/PaperAdapter.java]: This was added while I was testing storing the villager reputation list which was recommended by a user and I forgot to remove the import. There was a commit that was released after that initial that removes it.

File: src/main/java/net/coreprotect/bukkit/Bukkit_v1_19.java (hunk 1 above)
Comment (Gungadean) [reply src/main/java/net/coreprotect/bukkit/Bukkit_v1_19.java]: Created commit fixing this issue.

File: src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java (hunk 2 above)
Comment (Gungadean) [reply src/main/java/net/coreprotect/listener/block/BlockSpreadListener.java:31]: Really all the events should just have ignoreCancelled added to the annotation but I was trying to not make too many major changes from formatting.

**Comment by mrfloris** [comment]
I hope Intelli wlil soon come back to the Minecraft community and update the plugin for 1.19.2 so we can further our upgrade on the network and have support for sculk and such. 

Thank you all so far for trying to make a PR to help make that happen.

**Pull request merged**